File: 1376-time-needed-to-inform-all-employees/1376-time-needed-to-inform-all-employees.cpp
class Solution {
public:
    int backTrack(int src, vector<int>& time, const vector<int>& manager, const vector<int>& informTime){
        if(time[src] >= 0){
            return time[src];
        }
        
        int manId = manager[src];
        
        int mantime = backTrack(manId, time, manager, informTime);
        
        time[src] = mantime + informTime[manId];
        
        return time[src];
        
    }
                  
    int numOfMinutes(int n, int headID, vector<int>& manager, vector<int>& informTime) {
        vector<int> time(n, -1);
        
        time[headID] = 0;
        
        int mxTime = 0;
        
        for(int src = 0; src < n; src++){
            if(time[src] < 0 ){
                time[src] = backTrack(src, time, manager, informTime);
            }
            mxTime = max(mxTime, time[src]);
        }
        
        return mxTime;
        
    }
};

File: 1376-time-needed-to-inform-all-employees/1376-time-needed-to-inform-all-employees.py
class Solution:
    def numOfMinutes(self, n: int, headID: int, manager: List[int], informTime: List[int]) -> int:
        timeInformed = n * [10000000]
        
        # REMINDER: DO NOT DO n * [[ ]] as it creates n references to the SAME list
        
        # adj = [ list() for i in range(n)]
        
        adj = defaultdict(list)
        
        for emp, man in enumerate(manager):
            if man != -1:
                adj[man].append(emp)
        
        def dfs(src, time) -> None:
            timeInformed[src] = time
            for emp in adj[src]:
                dfs(emp, time + informTime[src])
      
        dfs(headID, 0)
        
        return max(timeInformed)
        